test/e2e/common/init_container.go

```diff
@@ -90,14 +99,23 @@ var _ = framework.KubeDescribe("InitContainer [NodeConformance]", func() {
 		}
 		e2elog.Logf("PodSpec: initContainers in spec.initContainers")
 		startedPod := podClient.Create(pod)
-		w, err := podClient.Watch(metav1.SingleObject(startedPod.ObjectMeta))
-		framework.ExpectNoError(err, "error watching a pod")
-		wr := watch.NewRecorder(w)
+
+		fieldSelector := fields.OneTermEqualSelector("metadata.name", startedPod.Name).String()
+		w := &cache.ListWatch{
+			WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
+				options.FieldSelector = fieldSelector
+				return podClient.Watch(options)
+			},
+		}
+		var events []watch.Event
 		ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), framework.PodStartTimeout)
 		defer cancel()
-		event, err := watchtools.UntilWithoutRetry(ctx, wr, conditions.PodCompleted)
+		event, err := watchtools.Until(ctx, startedPod.ResourceVersion, w,
+			recordEvents(events, conditions.PodCompleted),
```

Comment:
So I know this test was grandfathered in, but actually violates a bunch of requirements we listed here - 

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements 

namely on conditions and events.

Reply:
This came up during a recent conformance meeting in the context of a different test.  This reads like it violates those constraints, but those are really referring to the v1.Event resource and Condition field.  In this case, watch.Event is being used, which does have some guarantees about delivery, and conditions.PodCompleted is a function which looks at Pod.Status.Phase which is acceptable.